paxos_response_handler: carry effective replication map

bhalevy opened the pull request:
As `create_write_response_handler` on this path accepts
an `inet_address_vector_replica_set` that corresponds to the
effective_replication_map_ptr in the paxos_response_handler,
but currently, the function retrieves a new
effective_replication_map_ptr
that may not hold all the said endpoints.

Fixes scylladb/scylladb#15138

avikivity commented on service/storage_proxy.cc:
```diff
@@ -1216,6 +1216,10 @@ class paxos_response_handler : public enable_shared_from_this<paxos_response_han
     // adn returns true when quorum of such requests are accumulated
     bool learned(gms::inet_address ep);
+
+    locator::effective_replication_map_ptr ermp() const noexcept {
```
Please use a longer name then ermp.

bhalevy replied:
okay

avikivity commented:
Looks good. I wonder if we can mechanically validate that all call sites of get_effective_replication_map() are true roots.

bhalevy commented:
> Looks good. I wonder if we can mechanically validate that all call sites of get_effective_replication_map() are true roots.

When we start passing around node*s instead of endpoints it would be clear they need a valid topology object that holds them, so their root e_r_m will have to be carried along during their life time.

bhalevy commented:
btw batch queries also needs to carry the e_r_m similar to this PR and #15142. I hope to get to it tomorrow.

scylladb-promoter commented:
CI state **`ABORTED`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3225/

bhalevy commented:
In v2 ([3af1e50](https://github.com/scylladb/scylladb/commit/3af1e50f0c12fae63556dba541235b1350c5b4cd)):
- rebased
- renamed paxos_handler::ermp to get_effective_replication_map

gleb-cloudius reviewed (approved):
Small nit but otherwise LGTM.

bhalevy commented on service/storage_proxy.cc:
```diff
@@ -3115,16 +3119,14 @@ storage_proxy::create_write_response_handler(const std::tuple<lw_shared_ptr<paxo
 
 result<storage_proxy::response_id_type>
-storage_proxy::create_write_response_handler(const std::tuple<lw_shared_ptr<paxos::proposal>, schema_ptr, dht::token, inet_address_vector_replica_set>& meta,
+storage_proxy::create_write_response_handler(const std::tuple<shared_ptr<paxos_response_handler>, lw_shared_ptr<paxos::proposal>, schema_ptr, dht::token, inet_address_vector_replica_set>& meta,
```
fair enough.  will change

bhalevy commented:
In v3 ([4a2e367](https://github.com/scylladb/scylladb/commit/4a2e367e92fc48bda04952db3ab7c44e37470c40)):
- changed tuple ordering to match pre-existing `create_write_response_handler` variant that accepts paxos_response_handler right before token.

avikivity commented:
> > Looks good. I wonder if we can mechanically validate that all call sites of get_effective_replication_map() are true roots.
> 
> When we start passing around node*s instead of endpoints it would be clear they need a valid topology object that holds them, so their root e_r_m will have to be carried along during their life time.

It doesn't prevent some random callee from getting its own replication map down the line.

This is going to be difficult to debug since it will only very rarely cause trouble.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3244/

bhalevy commented:
@scylladb/scylla-maint please merge
